File: AlgorithmStudy/9주차/BOJ1012/김재환1012.java
import java.io.*;
import java.util.*;

public class Main {
    static int[] dr = {-1,1,0,0};
    static int[] dc = {0,0,-1,1};

    public static void main(String[] args)throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int T = Integer.parseInt(br.readLine().trim());
        while (T-- > 0) {
            st = new StringTokenizer(br.readLine());
            int M = Integer.parseInt(st.nextToken());
            int N = Integer.parseInt(st.nextToken());
            int K = Integer.parseInt(st.nextToken());

            // 배추밭 표시 (N행 M열)
            int[][] field = new int[N][M];
            for (int i = 0; i < K; i++) {
                st = new StringTokenizer(br.readLine());
                int x = Integer.parseInt(st.nextToken());
                int y = Integer.parseInt(st.nextToken());
                field[y][x] = 1;
            }

            boolean[][] visited = new boolean[N][M];
            int count = 0;

            for (int r = 0; r < N; r++) {
                for (int c = 0; c < M; c++) {
                    if (field[r][c] == 1 && !visited[r][c]) {
                        // BFS로 연결된 모든 배추 처리
                        Queue<int[]> queue = new ArrayDeque<>();
                        queue.add(new int[]{r, c});
                        visited[r][c] = true;

                        while (!queue.isEmpty()) {
                            int[] cur = queue.poll();
                            int cr = cur[0], cc = cur[1];
                            for (int d = 0; d < 4; d++) {
                                int nr = cr + dr[d];
                                int nc = cc + dc[d];
                                if (nr >= 0 && nr < N && nc >= 0 && nc < M) {
                                    if (field[nr][nc] == 1 && !visited[nr][nc]) {
                                        visited[nr][nc] = true;
                                        queue.add(new int[]{nr, nc});
                                    }
                                }
                            }
                        }
                        count++;
                    }
                }
            }

            System.out.println(count);
        }
    }
}

File: AlgorithmStudy/9주차/BOJ1012/김재환1012.py
import sys
from collections import deque

input = sys.stdin.readline

# 상하좌우 이동 방향
dr = [-1, 1, 0, 0]
dc = [0, 0, -1, 1]

T = int(input().strip())
for _ in range(T):
    # M: 가로 길이, N: 세로 길이, K: 배추 개수
    M, N, K = map(int, input().split())

    # 배추 위치 표시 배열: 0 = 없음, 1 = 있음
    field = [[0]*M for _ in range(N)]
    for __ in range(K):
        x, y = map(int, input().split())
        field[y][x] = 1

    # 방문 체크
    visited = [[False]*M for _ in range(N)]

    def bfs(sr, sc):
        queue = deque()
        queue.append((sr, sc))
        visited[sr][sc] = True

        while queue:
            r, c = queue.popleft()
            for i in range(4):
                nr = r + dr[i]
                nc = c + dc[i]
                # 범위 체크
                if 0 <= nr < N and 0 <= nc < M:
                    # 배추 있고 아직 방문 안 했으면
                    if field[nr][nc] == 1 and not visited[nr][nc]:
                        visited[nr][nc] = True
                        queue.append((nr, nc))

    # 군집 수 세기
    count = 0
    for r in range(N):
        for c in range(M):
            if field[r][c] == 1 and not visited[r][c]:
                # 새 연결 요소 발견
                bfs(r, c)
                count += 1

    print(count)
